wordpress-mobile/WordPress-iOS
# [Tooling] Add Swiftlint rule for NSLocalizedString -> AppLocalizedString

This work is part of parent project paaHJt-2Ib-p2 & subproject paaHJt-2J8-p2
Especially see comment paaHJt-2J8-p2#comment-5650

~~ℹ️  This PR is built on top of https://github.com/wordpress-mobile/WordPress-iOS/pull/17676 (target branch: `cleanup/project-files`) because some of the obsolete files that got deleted had violations; so I chose to delete those obsolete files as part of https://github.com/wordpress-mobile/WordPress-iOS/pull/17676 before fixing the remaining, valid files to use `AppLocalizedString`.~~ Update: #17676 has been merged, base is now the standard main branch.

## What it does

 - Add a `swiftlint` rule to enforce the use of `AppLocalizedString` instead of `NSLocalizedString` everywhere except in Tests and except in `.m` files (because top-level functions in Swift are not exposed to ObjC…)
 - Replace all occurrences of `NSLocalizedString` with `AppLocalizedString` in every `*.swift` files except tests
 - Fix one of the Unit Test which was relying on the locale the test was run on to be English
 - Fix the SwiftLint aggregate target / build phase to be compatible with `rbenv`

## RFC: Is it the right approach?

TBH I'm not sure this is the right approach for multiple reasons:

 - This is a bit unusual/unconventional to have to use `AppLocalizedString` and not `NSLocalizaedString` in every place – even the ones only used in the app
 - This enforcement will differ from other app codebases like WooCommerce, because WCiOS does not currently have app extensions or widgets, so there's no reason to enforce a similar rule on WCiOS. Which means we'd have different rules / conventions for localized string in WPiOS vs WCiOS

:one: IMHO, it would make way more sense to extract any source code that is common and used in both the app and the widget(s) and app extensions into one or more separate framework target (e.g WPStats.framework), then make the app and widget target both depend on it. That would then allow us to focus the SwiftLint rule about `AppLocalizedString` only on this target's folder — instead of enforcing it everywhere.

But that ☝️  would be a change out of scope of my current project, which is why I didn't go that route.

:two: Instead, this Draft PR is meant to be used as a point of discussion and see the impact that it could have if we still go with the "Enforce `AppLocalizedString` everywhere" solution instead in the meantime, its impacts and its Pros and Cons.

3️⃣  Another, middle-ground solution would be to keep the status quo (`NSLocalizedString` in most places, no separate fmk yet) and just specify the paths to the shared files explicitly in the `.swiftlint.yml`; but that would not be very scalable if more files are added to the app extension / widget targets…

## PR Status: Ready to Review / RFC, but still Draft / Do Not Merge

The PR is ready for review, but it was created more as a way to see the impact it would have on the codebase and start the discussion mentioned above.
Besides, even if we decide to go with that solution, I'd still want to write a P2 post to officially announce such a change before landing it.

## To test

 - Select the "SwiftLint" scheme in Xcode and build it (⌘B), and check that it passes without error nor warning. (Alternatively, you can run `bundle exec rake lint` from the Terminal.)
 - Introduce an intentional violation by replacing one of the `AppLocalizedString` instance in the code with `NSLocalizedString`, re-run the SwiftLint scheme, and check that it complains about the violation. Then revert your change.
 - Ensure that the project builds for both the WordPress and Jetpack schemes
 - Ensure the Unit Tests still build and pass
 - Smoke-test the app in a locale other than English to ensure that it still uses the translations for that non-English locale
